boards: nxp: frdm_mcxn947 and mcx_n9xx_evk: boot directly from QSPI flash

DerekSnell opened the pull request:
The ROM bootloader has the option to boot from external QSPI flash on the FlexSPI instead of internal flash.  Configures QSPI board variants to boot directly from QSPI.  Secondary bootloader is no longer required.  Moves the MCUboot boot_partition to QSPI flash.

Placing the MCUboot boot_partition in the QSPI flash depends on https://github.com/mcu-tools/mcuboot/pull/2319

Commit:
boards: nxp: frdm_mcxn947_cpu0_qspi: boot directly from QSPI flash

Configures QSPI board variant to boot directly from external QSPI flash
on the FlexSPI.  Secondary bootloader is no longer required.  Moves the
MCUboot boot_partition to QSPI flash.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
boards: nxp: mcx_n9xx_evk_cpu0_qspi: boot directly from QSPI flash

Configures QSPI board variant to boot directly from external QSPI flash
on the FlexSPI.  Secondary bootloader is no longer required.  Moves the
MCUboot boot_partition to QSPI flash.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

decsny commented:
DNM until https://github.com/zephyrproject-rtos/zephyr/pull/90566 merges, otherwise CI breaks, thanks

Commit:
soc: nxp: MCXNx4x: enable booting from QSPI flash

The ROM bootloader has the option to boot from external QSPI flash on
the FlexSPI instead of internal flash.  Adds
CONFIG_NXP_FLEXSPI_BOOT_HEADER to include the FlexSPI boot ROM header
in the image.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

DerekSnell commented:
I am documenting this here for those who prefer using the previous hybrid flash use-case.  This PR enabled booting directly from external QSPI flash.  Before this PR merged, XIP from QSPI was supported using both internal and external flash, what I call here the hybrid use-case.  This hybrid method boots from internal flash with the MCUboot boot_partition.  Then MCUboot enables the FlexSPI interface and configures for the external QSPI flash, and then loads the app to XIP from QSPI.  However, MCUboot was [reverted](https://github.com/mcu-tools/mcuboot/pull/2276) which broke this hybrid use-case.  But the hybrid option is no longer needed, because now the QSPI board variant will boot directly from QSPI, and no longer requires MCUboot.

If the hybrid flash use-case is preferred, these steps work for me.  These steps are written for the `frdm_mcxn947` board.  MCUboot is built and flashed in internal flash.  The blinky app is built for the `frdm_mcxn947//cpu0/qspi` board variant and flashed to the external QSPI.  These steps are based on https://github.com/zephyrproject-rtos/zephyr/commit/c7e36f01696a206f4e0db4452602f76b31c19d09 :

1. Program the PFR to boot from internal flash.  This is the default setting of the MCXN947 SOC.  But if the SOC was previously programmed to boot from external QSPI flash, it needs to be programmed back.  For more details, see [Programming the ROM bootloader for external QSPI](https://docs.zephyrproject.org/latest/boards/nxp/frdm_mcxn947/doc/index.html#programming-the-rom-bootloader-for-external-qspi).
2. In the MCUboot repo, revert https://github.com/mcu-tools/mcuboot/commit/81315483fcbdf1f1524c2b34a1fd4de6c77cd0f4
3. Apply the attached patch [mcuboot-hybrid-flash-overlay.patch.txt](https://github.com/user-attachments/files/20510624/mcuboot-hybrid-flash-overlay.patch.txt) to the Zephyr repo.  This adds mcuboot-hybrid-flash.overlay to the FRDM_MCXN947 board folder.
4. Build and flash MCUboot and the blinky sample using the commands below:
```
west build -b frdm_mcxn947//cpu0 ../bootloader/mcuboot/boot/zephyr -d ../hybrid_mcb_n947 --pristine -- -DDTC_OVERLAY_FILE="../../../../zephyr/boards/nxp/frdm_mcxn947/mcuboot-hybrid-flash.overlay" 
west build -b frdm_mcxn947//cpu0/qspi samples/basic/blinky -d ../hybrid_blinky_n947 --pristine -- -DCONFIG_BOOTLOADER_MCUBOOT=y -DDTC_OVERLAY_FILE="../../boards/nxp/frdm_mcxn947/mcuboot-hybrid-flash.overlay" -DCONFIG_MCUBOOT_SIGNATURE_KEY_FILE=\"bootloader/mcuboot/root-rsa-2048.pem\" 
west flash -d ../hybrid_mcb_n947 -r jlink
west flash -d ../hybrid_blinky_n947 -r jlink
```

The debug console output shows MCUboot boots and loads the blinky sample:
```
*** Booting MCUboot v2.2.0-rc1-17-gbcd9614727ab ***
*** Using Zephyr OS build v4.1.0-4901-gc7e36f01696a ***
I: Starting bootloader
I: Image index: 0, Swap type: none
I: Bootloader chainload address offset: 0x0
I: Image version: v0.0.0
I: Jumping to the first image slot
*** Booting Zephyr OS build v4.1.0-4901-gc7e36f01696a ***
LED state: OFF
LED state: ON
```

butok commented:
> for those who prefer using the previous hybrid flash use-case.

Known issue: For the previous "hybrid flash use-case", the MCU-Boot firmware update did not work for me.
